Sources/TuistLoader/Utils/ManifestFilesLocator.swift

```diff
@@ -73,18 +73,46 @@ public final class ManifestFilesLocator: ManifestFilesLocating {
         } else {
             let path = rootDirectoryLocator.locate(from: locatingPath) ?? locatingPath
 
-            let excludedPaths = excluding
-                .flatMap {
-                    FileHandler.shared.glob(path, glob: $0)
+            let pluginsPaths = fetchTuistManifestsFilePaths(at: path)
+                .filter { $0.basename == Manifest.plugin.fileName(path) }
+                .filter { path in
+                    !excluding.contains { pattern in match(path, pattern: pattern) }
                 }
 
-            let pluginsPaths = Set(FileHandler.shared.glob(path, glob: "**/\(Manifest.plugin.fileName(path))"))
-                .subtracting(excludedPaths)
-
             return Array(pluginsPaths)
         }
     }
 
+    private func match(_ path: AbsolutePath, pattern: String) -> Bool {
+        return fnmatch(pattern, path.pathString, 0) != FNM_NOMATCH
+    }
+
+    var cacheTuistManifestsFilePaths = [AbsolutePath: Set<AbsolutePath>]()
+
+    private func fetchTuistManifestsFilePaths(at path: AbsolutePath) -> Set<AbsolutePath> {
+        if let cachedTuistManifestsFilePaths = cacheTuistManifestsFilePaths[path] {
+            return cachedTuistManifestsFilePaths
+        }
+        let fileNamesCandidates: Set<String> = [
+            Manifest.project.fileName(path),
+            Manifest.workspace.fileName(path),
+            Manifest.plugin.fileName(path),
+        ]
+        let tuistManifestsFilePaths = FileHandler.shared.glob(path, glob: "**/*.swift")
+            .filter { fileNamesCandidates.contains($0.basename) }
+            .filter { hasValidManifestContent($0) }
+        let cachedTuistManifestsFilePaths = Set(tuistManifestsFilePaths)
+        cacheTuistManifestsFilePaths[path] = cachedTuistManifestsFilePaths
+        return cachedTuistManifestsFilePaths
+    }
+
+    private func hasValidManifestContent(_ path: AbsolutePath) -> Bool {
+        guard let content = try? FileHandler.shared.readTextFile(path) else { return false }
```

Comment:
Would it be better to do this lazily by streaming the file and searching the first couple lines for the import?

Reply:
That would work for very big files, but here there won't be any performance improvements, that's for two reasons:
* we are checking the content for the few files which MAY be a manifest file (Project.swift etc...) and not all the source files. 
* even if we would start stream lines, that is usually implemented by loading a buffer (a couple of mega) and returning that as line by line. That means that in our scenario, where a manifest is way under this buffer, we would still load the whole file, getting no concrete improvements

Anyway, the implementation of this PR it's quite fast in fact it now performs faster that the current implementation in main branch (-75%) ☺️

Reply:
Amazing! Makes sense thanks for the details.